Sale Price Prediction Models (Person B)

This notebook focuses on predicting house sale prices using:
- Linear Regression
- Regularization (Ridge, LASSO)
- Dimension Reduction (PCR, PLS)
- Nonlinear Models
- Cross Validation

Evaluation Metrics:
- RMSE
- MSE
- R²
- Cross-validation


In [ ]:
from functools import partial
print = partial(print, flush=True)

In [ ]:
import os
import shutil

mountpoint = '/content/drive'

# Check if the mountpoint directory exists and is not empty
if os.path.isdir(mountpoint) and os.listdir(mountpoint):
    print(f"Clearing existing files/directories in {mountpoint}...")
    for item in os.listdir(mountpoint):
        item_path = os.path.join(mountpoint, item)
        try:
            if os.path.isfile(item_path) or os.path.islink(item_path):
                os.remove(item_path)
            elif os.path.isdir(item_path):
                shutil.rmtree(item_path)
            print(f"Removed: {item_path}")
        except Exception as e:
            print(f"Error removing {item_path}: {e}")
else:
    print(f"Mountpoint {mountpoint} is already empty or does not exist.")

# Now attempt to mount Google Drive
from google.colab import drive
drive.mount(mountpoint)

Mountpoint /content/drive is already empty or does not exist.
Mounted at /content/drive


In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [ ]:
cd /content/drive/MyDrive/Stat_4210/FINAL_PROJECT/

/content/drive/MyDrive/Stat_4210/FINAL_PROJECT


In [ ]:
import pandas as pd

df = pd.read_csv("AmesHousing.csv")
df.head()

,Order,PID,MS SubClass,MS Zoning,Lot Frontage,Lot Area,Street,Alley,Lot Shape,Land Contour,...,Pool Area,Pool QC,Fence,Misc Feature,Misc Val,Mo Sold,Yr Sold,Sale Type,Sale Condition,SalePrice
0,1,526301100,20,RL,141.0,31770,Pave,NaN,IR1,Lvl,...,0,NaN,NaN,NaN,0,5,2010,WD,Normal,215000
1,2,526350040,20,RH,80.0,11622,Pave,NaN,Reg,Lvl,...,0,NaN,MnPrv,NaN,0,6,2010,WD,Normal,105000
2,3,526351010,20,RL,81.0,14267,Pave,NaN,IR1,Lvl,...,0,NaN,NaN,Gar2,12500,6,2010,WD,Normal,172000
3,4,526353030,20,RL,93.0,11160,Pave,NaN,Reg,Lvl,...,0,NaN,NaN,NaN,0,4,2010,WD,Normal,244000
4,5,527105010,60,RL,74.0,13830,Pave,NaN,IR1,Lvl,...,0,NaN,MnPrv,NaN,0,3,2010,WD,Normal,189900


In [ ]:
df.shape
df.info()
df.describe()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2930 entries, 0 to 2929
Data columns (total 82 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Order            2930 non-null   int64  
 1   PID              2930 non-null   int64  
 2   MS SubClass      2930 non-null   int64  
 3   MS Zoning        2930 non-null   object 
 4   Lot Frontage     2440 non-null   float64
 5   Lot Area         2930 non-null   int64  
 6   Street           2930 non-null   object 
 7   Alley            198 non-null    object 
 8   Lot Shape        2930 non-null   object 
 9   Land Contour     2930 non-null   object 
 10  Utilities        2930 non-null   object 
 11  Lot Config       2930 non-null   object 
 12  Land Slope       2930 non-null   object 
 13  Neighborhood     2930 non-null   object 
 14  Condition 1      2930 non-null   object 
 15  Condition 2      2930 non-null   object 
 16  Bldg Type        2930 non-null   object 
 17  House Style   

,Order,PID,MS SubClass,Lot Frontage,Lot Area,Overall Qual,Overall Cond,Year Built,Year Remod/Add,Mas Vnr Area,...,Wood Deck SF,Open Porch SF,Enclosed Porch,3Ssn Porch,Screen Porch,Pool Area,Misc Val,Mo Sold,Yr Sold,SalePrice
count,2930.00000,2.930000e+03,2930.000000,2440.000000,2930.000000,2930.000000,2930.000000,2930.000000,2930.000000,2907.000000,...,2930.000000,2930.000000,2930.000000,2930.000000,2930.000000,2930.000000,2930.000000,2930.000000,2930.000000,2930.000000
mean,1465.50000,7.144645e+08,57.387372,69.224590,10147.921843,6.094881,5.563140,1971.356314,1984.266553,101.896801,...,93.751877,47.533447,23.011604,2.592491,16.002048,2.243345,50.635154,6.216041,2007.790444,180796.060068
std,845.96247,1.887308e+08,42.638025,23.365335,7880.017759,1.411026,1.111537,30.245361,20.860286,179.112611,...,126.361562,67.483400,64.139059,25.141331,56.087370,35.597181,566.344288,2.714492,1.316613,79886.692357
min,1.00000,5.263011e+08,20.000000,21.000000,1300.000000,1.000000,1.000000,1872.000000,1950.000000,0.000000,...,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,1.000000,2006.000000,12789.000000
25%,733.25000,5.284770e+08,20.000000,58.000000,7440.250000,5.000000,5.000000,1954.000000,1965.000000,0.000000,...,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,4.000000,2007.000000,129500.000000
50%,1465.50000,5.354536e+08,50.000000,68.000000,9436.500000,6.000000,5.000000,1973.000000,1993.000000,0.000000,...,0.000000,27.000000,0.000000,0.000000,0.000000,0.000000,0.000000,6.000000,2008.000000,160000.000000
75%,2197.75000,9.071811e+08,70.000000,80.000000,11555.250000,7.000000,6.000000,2001.000000,2004.000000,164.000000,...,168.000000,70.000000,0.000000,0.000000,0.000000,0.000000,0.000000,8.000000,2009.000000,213500.000000
max,2930.00000,1.007100e+09,190.000000,313.000000,215245.000000,10.000000,9.000000,2010.000000,2010.000000,1600.000000,...,1424.000000,742.000000,1012.000000,508.000000,576.000000,800.000000,17000.000000,12.000000,2010.000000,755000.000000


In [ ]:
important_vars = ["SalePrice", "Gr Liv Area", "Overall Qual", "Year Built"]
df[important_vars].describe()

df["SalePrice"].quantile([0.25, 0.5, 0.75])

,SalePrice
0.25,129500.0
0.50,160000.0
0.75,213500.0


In [ ]:
df = df.drop(["Order", "PID"], axis=1)

In [ ]:
num_cols = df.select_dtypes(include=["int64", "float64"]).columns
cat_cols = df.select_dtypes(include=["object"]).columns

In [ ]:
for col in num_cols:
    df[col] = df[col].fillna(df[col].median())

for col in cat_cols:
    df[col] = df[col].fillna("None")

In [ ]:
df = pd.get_dummies(df, drop_first=True)

In [ ]:
X = df.drop("SalePrice", axis=1)
y = df["SalePrice"]

In [ ]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=1
)

In [ ]:
from sklearn.preprocessing import StandardScaler

scaler = StandardScaler()

X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)

In [ ]:
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error, r2_score

lin_model = LinearRegression()
lin_model.fit(X_train, y_train)

y_pred_lin = lin_model.predict(X_test)

mse_lin = mean_squared_error(y_test, y_pred_lin)
rmse_lin = mse_lin ** 0.5
r2_lin = r2_score(y_test, y_pred_lin)

print("Linear RMSE:", rmse_lin)
print("Linear R2:", r2_lin)

Linear RMSE: 21348.906306283585
Linear R2: 0.9265106325394219


In [ ]:
import statsmodels.api as sm
import pandas as pd

X_train_sm = sm.add_constant(X_train)

# Ensure all columns are numeric, coercing errors to NaN
# This will handle any 'object' dtype columns that contain non-numeric strings
# or mixed types by turning non-convertible entries into NaN.
for col in X_train_sm.columns:
    X_train_sm[col] = pd.to_numeric(X_train_sm[col], errors='coerce')

# Drop any columns that became entirely NaN after coercion (i.e., contained no numeric data)
X_train_sm = X_train_sm.dropna(axis=1, how='all')

# Fill any remaining NaNs (introduced by coercion for individual non-numeric values)
# For simplicity, filling with 0. A more sophisticated imputation could be used if needed.
X_train_sm = X_train_sm.fillna(0)

# Explicitly convert X_train_sm and y_train to NumPy arrays of float type
# This ensures that statsmodels receives purely numeric data.
X_train_sm_numeric = X_train_sm.values.astype(float)
y_train_numeric = y_train.values.astype(float)

ols_model = sm.OLS(y_train_numeric, X_train_sm_numeric).fit()

print("AIC:", ols_model.aic)
print("BIC:", ols_model.bic)

AIC: 53763.61419620982
BIC: 55324.469631048676


In [ ]:
from sklearn.linear_model import RidgeCV
import numpy as np

ridge_model = RidgeCV(
    alphas=np.logspace(-3,3,100),
    cv=5
)

ridge_model.fit(X_train_scaled, y_train)

y_pred_ridge = ridge_model.predict(X_test_scaled)

rmse_ridge = np.sqrt(mean_squared_error(y_test, y_pred_ridge))
r2_ridge = r2_score(y_test, y_pred_ridge)

print("Best Ridge Alpha:", ridge_model.alpha_)

Best Ridge Alpha: 432.87612810830615


In [ ]:
from sklearn.linear_model import LassoCV

lasso_model = LassoCV(
    cv=5,
    max_iter=10000
)

lasso_model.fit(X_train_scaled, y_train)

y_pred_lasso = lasso_model.predict(X_test_scaled)

rmse_lasso = np.sqrt(mean_squared_error(y_test, y_pred_lasso))
r2_lasso = r2_score(y_test, y_pred_lasso)

print("Best LASSO Alpha:", lasso_model.alpha_)
print("Nonzero Coefs:", sum(lasso_model.coef_ != 0))

Best LASSO Alpha: 1115.7273163741354
Nonzero Coefs: 80


In [ ]:
from sklearn.preprocessing import PolynomialFeatures

poly = PolynomialFeatures(degree=2)

X_train_poly = poly.fit_transform(X_train)
X_test_poly = poly.transform(X_test)

poly_model = LinearRegression()
poly_model.fit(X_train_poly, y_train)

y_pred_poly = poly_model.predict(X_test_poly)

mse_poly = mean_squared_error(y_test, y_pred_poly)
rmse_poly = mse_poly ** 0.5
r2_poly = r2_score(y_test, y_pred_poly)

In [ ]:
from sklearn.decomposition import PCA

best_rmse = 999999
best_comp = 0

for n in range(5,31):

    pca = PCA(n_components=n)
    X_train_pca = pca.fit_transform(X_train_scaled)
    X_test_pca = pca.transform(X_test_scaled)

    model = LinearRegression()
    model.fit(X_train_pca, y_train)

    pred = model.predict(X_test_pca)

    rmse = np.sqrt(mean_squared_error(y_test,pred))

    if rmse < best_rmse:
        best_rmse = rmse
        best_comp = n

rmse_pcr = best_rmse
print("Best PCR Components:", best_comp)

Best PCR Components: 30


In [ ]:
from sklearn.cross_decomposition import PLSRegression

best_rmse = 999999
best_comp = 0

for n in range(1,21):

    pls = PLSRegression(n_components=n)
    pls.fit(X_train_scaled, y_train)

    pred = pls.predict(X_test_scaled)

    rmse = np.sqrt(mean_squared_error(y_test,pred))

    if rmse < best_rmse:
        best_rmse = rmse
        best_comp = n

rmse_pls = best_rmse
print("Best PLS Components:", best_comp)

Best PLS Components: 7


In [ ]:
!pip install pygam

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 84.6/84.6 kB 3.3 MB/s eta 0:00:00


In [ ]:
from pygam import LinearGAM, s

X_train_gam = X_train[['Gr Liv Area','Overall Qual','Year Built']]
X_test_gam = X_test[['Gr Liv Area','Overall Qual','Year Built']]

gam = LinearGAM(
    s(0)+s(1)+s(2)
)

gam.fit(X_train_gam, y_train)

pred_gam = gam.predict(X_test_gam)

rmse_gam = np.sqrt(mean_squared_error(y_test,pred_gam))

In [ ]:
from sklearn.model_selection import cross_val_score

cv_rmse_lin = np.mean(np.sqrt(-cross_val_score(
    lin_model, X, y,
    scoring="neg_mean_squared_error",
    cv=5
)))

print("Linear CV RMSE:", cv_rmse_lin)

Linear CV RMSE: 35251.70885364136


In [ ]:
results = pd.DataFrame({
    "Model":[
        "Linear",
        "Ridge",
        "LASSO",
        "Polynomial",
        "PCR",
        "PLS",
        "GAM"
    ],
    "RMSE":[
        rmse_lin,
        rmse_ridge,
        rmse_lasso,
        rmse_poly,
        rmse_pcr,
        rmse_pls,
        rmse_gam
    ]
})

print(results.sort_values("RMSE"))

        Model           RMSE
0      Linear   21348.906306
1       Ridge   22600.061089
5         PLS   23453.882508
2       LASSO   23602.074047
4         PCR   29827.511874
6         GAM   30247.210888
3  Polynomial  397743.667912


In [ ]:
from sklearn.model_selection import cross_val_score
import numpy as np

cv_rmse = np.sqrt(-cross_val_score(
    lin_model, X, y,
    scoring="neg_mean_squared_error",
    cv=5
))

print("CV RMSE:", cv_rmse.mean())

CV RMSE: 35251.70885364136


In [ ]:
print("CV-MSE was used as an estimate of Expected Prediction Error (EPE).")
print("Models were compared using RMSE, CV-RMSE, AIC, and BIC.")

CV-MSE was used as an estimate of Expected Prediction Error (EPE).
Models were compared using RMSE, CV-RMSE, AIC, and BIC.


## Model Performance Comparison

Across all models tested, the standard linear regression model performed best, achieving the lowest RMSE of approximately 21,349 and the highest predictive accuracy on the test set. This suggests that the Ames Housing dataset contains a strong linear relationship between predictors and sale price.

The Ridge regression model performed second best with an RMSE of approximately 22,600, indicating that coefficient shrinkage helped maintain strong predictive performance while reducing potential overfitting.

The PLS regression model also performed well with an RMSE of approximately 23,454, outperforming both LASSO and PCR. This suggests that reducing predictors into latent components while incorporating the response variable can be useful for housing price prediction.

The LASSO regression model had an RMSE of approximately 23,602. Although it did not outperform linear regression, it remained competitive while shrinking many coefficients to zero, making it useful for variable selection and simpler interpretation.

## Effect of Regularization

Both Ridge and LASSO regression were used to penalize large coefficients and reduce overfitting.

Ridge kept all predictors while shrinking coefficient size.
LASSO selected a smaller subset of variables by forcing some coefficients to zero.

In this dataset, regularization did not outperform the baseline linear model, suggesting that the original predictors already captured the relationship effectively.

## Dimension Reduction Results

Principal Components Regression (PCR) achieved an RMSE of approximately 29,828, while Partial Least Squares (PLS) performed much better with an RMSE of 23,454.

This indicates that PLS was more effective because it forms components using both predictors and the response variable, whereas PCR only uses predictor variation.

## Nonlinear Model Performance

The GAM model achieved an RMSE of approximately 30,247, showing moderate predictive ability but not outperforming simpler linear methods.

The Polynomial regression model performed extremely poorly with an RMSE near 397,744, indicating severe overfitting caused by excessive model complexity.

These results suggest that nonlinear methods were unnecessary for this dataset.

## Cross-Validation Results

The 5-fold cross-validation RMSE for the linear model was approximately 35,252, which was higher than the single train-test split RMSE.

This suggests some variation across folds, but overall the linear model still generalized reasonably well.

AIC / BIC Results

AIC and BIC were calculated for the linear regression model:

AIC = 53,763
BIC = 55,324

These metrics help compare model fit while penalizing model complexity. Lower values indicate a better balance of fit and simplicity.

## Overall Conclusion

Overall, the results show that simple linear regression was the best predictive model for the Ames Housing dataset.

More complex methods such as GAM, PCR, and polynomial regression did not improve performance. Ridge, LASSO, and PLS remained competitive but did not exceed the baseline model.

This suggests that the dataset has a strong linear structure, making traditional regression methods highly effective for sale price prediction.

In [ ]:
!pip -q install -U nbconvert[webpdf]

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 46.2/46.2 MB 9.5 MB/s eta 0:00:00


In [ ]:
!python -m playwright install --with-deps chromium > /dev/null

W: Skipping acquire of configured file 'main/source/Sources' as repository 'https://r2u.stat.illinois.edu/ubuntu jammy InRelease' does not seem to provide it (sources.list entry misspelt?)
(node:25544) [DEP0169] DeprecationWarning: `url.parse()` behavior is not standardized and prone to errors that have security implications. Use the WHATWG URL API instead. CVEs are not issued for `url.parse()` vulnerabilities.
(Use `node --trace-deprecation ...` to show where the warning was created)
(node:25583) [DEP0169] DeprecationWarning: `url.parse()` behavior is not standardized and prone to errors that have security implications. Use the WHATWG URL API instead. CVEs are not issued for `url.parse()` vulnerabilities.
(Use `node --trace-deprecation ...` to show where the warning was created)
(node:25598) [DEP0169] DeprecationWarning: `url.parse()` behavior is not standardized and prone to errors that have security implications. Use the WHATWG URL API instead. CVEs are not issued for `url.parse()` 

In [ ]:
!jupyter nbconvert --to webpdf "/content/drive/MyDrive/Stat_4210/FINAL_PROJECT/Stat4210_Project.ipynb" --output-dir "/content/drive/MyDrive/Stat_4210/FINAL_PROJECT"

[NbConvertApp] Converting notebook /content/drive/MyDrive/Stat_4210/FINAL_PROJECT/Stat4210_Project.ipynb to webpdf
[NbConvertApp] Building PDF
[NbConvertApp] PDF successfully created
[NbConvertApp] Writing 67158 bytes to /content/drive/MyDrive/Stat_4210/FINAL_PROJECT/Stat4210_Project.pdf
